File: src/wine/system/io/DevTTY.java
package wine.system.io;

import wine.system.kernel.Syscall;
import wine.emulation.Memory;
import wine.system.WineThread;
import wine.util.Log;

public class DevTTY implements FSNodeAccess {
    static public int active = 0;

    static public final int VT_AUTO = 0;
    static public final int VT_PROCESS = 1;
    static public final int VT_ACKACQ = 2;

    public int mode = VT_AUTO;

    static public final int K_RAW = 0x00;
    static public final int K_XLATE = 0x01;
    static public final int K_MEDIUMRAW = 0x02;
    static public final int K_UNICODE = 0x03;
    static public final int KDSKBMODE = 0x04;

    public int kbMode = K_UNICODE;

    public int waitv;
    public int relsig;
    public int acqsig;
    public boolean graphics = false;
    public Termios termios = new Termios();

    public DevTTY(int index) {
    }

    public boolean setLength(long length) {
        return false;
    }

    public boolean open(String mode) {
        return true;
    }
    public long getFilePointer() {
        return 0;
    }

    public boolean seek(long pos) {
        return false;
    }

    public int read(byte[] b) {
        return 0;
    }

    public boolean write(byte[] b) {
        StringBuilder builder = new StringBuilder();
        for (int i=0;i<b.length;i++) {
            builder.append((char) (b[i] & 0xFF));
        }
        WineThread.getCurrent().out(builder.toString());
        return true;
    }

    public void close() {
    }

    public int ioctl(WineThread thread, int request, Syscall.SyscallGetter getter) {
        Memory memory = thread.process.memory;
        switch (request) {
            case 0x4B32: // KBSETLED
                break;
            case 0x4B3A: // KDSETMODE
                graphics = getter.next()==1;
                break;
            case 0x4B44: // KDGKBMODE
                memory.writed(getter.next(), kbMode);
                break;
            case 0x4B45: // KDSKBMODE
                kbMode = getter.next();
                break;
            case 0x4B46: { // KDGKBENT
                int kbentry = getter.next();
                int table = memory.readb(kbentry);
                int index = memory.readb(kbentry+1);
                int value = memory.readw(kbentry+2);
                switch (table) {
                    case 0: // K_NORMTAB
                        memory.writew(kbentry+2, index);
                        break;
                    case 1: // K_SHIFTTAB
                        memory.writew(kbentry + 2, Character.toUpperCase((char) index));
                        break;
                    case 2: // K_ALTTAB
                        memory.writew(kbentry+2, index);
                        break;
                    case 3: // K_ALTSHIFTTAB
                        memory.writew(kbentry+2, index);
                        break;
                    default:
                        memory.writew(kbentry+2, index);
                        break;
                }
                break;
            }
            case 0x4B51: // KDSKBMUTE
                return -1;
            case 0x5401: // TCGETS
                termios.write(memory, getter.next());
                break;
            case 0x5402: // TCSETS
                termios.read(memory, getter.next());
                break;
            case 0x5600: // VT_OPENQRY
                memory.writed(getter.next(), 2);
                break;
            case 0x5601: { // VT_GETMODE
                int address = getter.next();
                memory.writeb(address, mode);
                memory.writeb(address+1, waitv); // waitv
                memory.writew(address+2, relsig); // relsig
                memory.writew(address+4, acqsig); // acqsig
                memory.writew(address+6, 0); // frsig
                break;
            }
            case 0x5602: { // VT_SETMODE
                int address = getter.next();
                mode = memory.readb(address); // VT_AUTO
                waitv = memory.readb(address+1); // waitv
                relsig = memory.readw(address+2); // relsig
                acqsig = memory.readw(address+4); // acqsig
                break;
            }
            case 0x5603: { // VT_GETSTATE
                int address = getter.next();
                memory.writew(address, 0); // v_active
                memory.writew(address, 0); // v_signal
                memory.writew(address, 1); // v_state
                break;
            }
            case 0x5605: { // VT_RELDISP
                break;
            }
            case 0x5606: // VT_ACTIVATE
                active = getter.next();
                break;
            case 0x5607: { // VT_WAITACTIVE
                int id = getter.next();
                if (id!=active)
                    Log.panic("VT_WAITACTIVE not fully implemented");
                break;
            }
            case 0x5608: { // VT_GETMODE
                break;
            }
            default:
                return -1;
        }
        return 0;
    }

    public int map(Memory memory, FileDescriptor fd, long off, int address, int len, boolean fixed, boolean read, boolean exec, boolean write, boolean shared) {
        Log.panic("Mapping /dev/tty not supported");
        return -1;
    }

    public boolean isReadReady() {
        return true;
    }

    public boolean isWriteReady() {
        return true;
    }
}

File: c/source/kernel/devs/devtty.c
/*
 *  Copyright (C) 2016  The BoxedWine Team
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
 */

#include "kerror.h"
#include "kthread.h"
#include "kprocess.h"
#include "kalloc.h"
#include "log.h"
#include "memory.h"
#include "fsapi.h"

#include <stdio.h>
#include <string.h>
#include <ctype.h>

#define VT_AUTO 0
#define VT_PROCESS 1
#define VT_ACKACQ 2   

#define K_RAW 0x00
#define K_XLATE 0x01
#define K_MEDIUMRAW 0x02
#define K_UNICODE 0x03
#define KDSKBMODE 0x04

static U32 activeTTY;

struct TTYData {
    U32 c_iflag;               /* input mode flags */
    U32 c_oflag;               /* output mode flags */
    U32 c_cflag;               /* control mode flags */
    U32 c_lflag;               /* local mode flags */
    U32 c_line;                    /* line discipline */
    U32 c_cc[19];                /* control characters */

    U32 mode;
    U32 kbMode;

    U32 waitv;
    U32 relsig;
    U32 acqsig;
    BOOL graphics;
};

void readTermios(struct Memory* memory, U32 address, struct TTYData* data) {
    int i;
    data->c_iflag = readd(memory, address);
    data->c_oflag = readd(memory, address+4);
    data->c_cflag = readd(memory, address+8);
    data->c_lflag = readd(memory, address+12);
    data->c_line = readb(memory, address+16);
    for (i=0;i<19;i++) {
        data->c_cc[i] = readb(memory, address+17+i);
    }
}

void writeTermios(struct Memory* memory, U32 address, struct TTYData* data) {
    int i;

    writed(memory, address, data->c_iflag);
    writed(memory, address+4, data->c_oflag);
    writed(memory, address+8, data->c_cflag);
    writed(memory, address+12, data->c_lflag);
    writeb(memory, address+16, data->c_line);
    for (i=0;i<19;i++) {
        writeb(memory, address+17+i, data->c_cc[i]);
    }
}

BOOL tty_init(struct KProcess* process, struct FsOpenNode* node) {
    struct TTYData* data = (struct TTYData*)kalloc(sizeof(struct TTYData), KALLOC_TTYDATA);
    data->mode = VT_AUTO;
    data->kbMode = K_UNICODE;
    node->data = data;
    return TRUE;
}

S64 tty_length(struct FsOpenNode* node) {
    return 0;
}

BOOL tty_setLength(struct FsOpenNode* node, S64 len) {
    return FALSE;
}

S64 tty_getFilePointer(struct FsOpenNode* node) {
    return 0;
}

S64 tty_seek(struct FsOpenNode* node, S64 pos) {
    return 0;
}

U32 tty_read(struct Memory* memory, struct FsOpenNode* node, U32 address, U32 len) {
    return 0;
}

static char buffer[PAGE_SIZE+1];

U32 tty_write(struct Memory* memory, struct FsOpenNode* node, U32 address, U32 len) {
    if (PAGE_SIZE-(address & (PAGE_SIZE-1)) >= len) {
        U8* ram = getPhysicalAddress(memory, address);
        memcpy(buffer, ram, len);
        buffer[len]=0;
        fprintf(stdout, "%s", buffer);		
        return len;
    } else {	
        U32 result = 0;
        while (len) {
            U32 todo = PAGE_SIZE-(address & (PAGE_SIZE-1));
            U8* ram = getPhysicalAddress(memory, address);
            if (todo>len)
                todo = len;
            memcpy(buffer, ram, todo);
            buffer[todo]=0;
            fprintf(stdout, "%s", buffer);
            len-=todo;
            address+=todo;
            result+=todo;
        }
        return result;
    }
}

void tty_close(struct FsOpenNode* node) {
    node->func->free(node);
}

U32 tty_ioctl(struct KThread* thread, struct FsOpenNode* node, U32 request) {
    struct TTYData* data = (struct TTYData*)node->data;
    struct CPU* cpu = &thread->cpu;

    switch (request) {
        case 0x4B32: // KBSETLED
            break;
        case 0x4B3A: // KDSETMODE
            data->graphics = IOCTL_ARG1==1;
            break;
        case 0x4B44: // KDGKBMODE
            writed(thread->process->memory, IOCTL_ARG1, data->kbMode);
            break;
        case 0x4B45: // KDSKBMODE
            data->kbMode = IOCTL_ARG1;
            break;
        case 0x4B46: { // KDGKBENT
            U32 kbentry = IOCTL_ARG1;
            U32 table = readb(thread->process->memory, kbentry);
            U32 index = readb(thread->process->memory, kbentry + 1);
            //U32 value = readw(thread->process->memory, kbentry+2);
            switch (table) {
                case 0: // K_NORMTAB
                    writew(thread->process->memory, kbentry + 2, index);
                    break;
                case 1: // K_SHIFTTAB
                    writew(thread->process->memory, kbentry + 2, toupper((char)index));
                    break;
                case 2: // K_ALTTAB
                    writew(thread->process->memory, kbentry + 2, index);
                    break;
                case 3: // K_ALTSHIFTTAB
                    writew(thread->process->memory, kbentry + 2, index);
                    break;
                default:
                    writew(thread->process->memory, kbentry + 2, index);
                    break;
            }
            break;
        }
        case 0x4B51: // KDSKBMUTE
            return -1;
        case 0x5401: // TCGETS
            writeTermios(thread->process->memory, IOCTL_ARG1, data);
            break;
        case 0x5402: // TCSETS
            readTermios(thread->process->memory, IOCTL_ARG1, data);
            break;
        case 0x5600: // VT_OPENQRY
            writed(thread->process->memory, IOCTL_ARG1, 2);
            break;
        case 0x5601: { // VT_GETMODE
            U32 address = IOCTL_ARG1;
            writeb(thread->process->memory, address, data->mode);
            writeb(thread->process->memory, address + 1, data->waitv); // waitv
            writew(thread->process->memory, address + 2, data->relsig); // relsig
            writew(thread->process->memory, address + 4, data->acqsig); // acqsig
            writew(thread->process->memory, address + 6, 0); // frsig
            break;
        }
        case 0x5602: { // VT_SETMODE
            U32 address = IOCTL_ARG1;
            data->mode = readb(thread->process->memory, address); // VT_AUTO
            data->waitv = readb(thread->process->memory, address + 1); // waitv
            data->relsig = readw(thread->process->memory, address + 2); // relsig
            data->acqsig = readw(thread->process->memory, address + 4); // acqsig
            break;
        }
        case 0x5603: { // VT_GETSTATE
            U32 address = IOCTL_ARG1;
            writew(thread->process->memory, address, 0); // v_active
            writew(thread->process->memory, address, 0); // v_signal
            writew(thread->process->memory, address, 1); // v_state
            break;
        }
        case 0x5605: { // VT_RELDISP
            break;
        }
        case 0x5606: // VT_ACTIVATE
            activeTTY = IOCTL_ARG1;
            break;
        case 0x5607: { // VT_WAITACTIVE
            U32 id = IOCTL_ARG1;
            if (id!=activeTTY)
                kpanic("VT_WAITACTIVE not fully implemented");
            break;
        }
        case 0x5608: { // VT_GETMODE
            break;
        }
        default:
            return -1;
    }
    return 0;
}

void tty_setAsync(struct FsOpenNode* node, struct KProcess* process, FD fd, BOOL isAsync) {
    if (isAsync)
        kwarn("tty_setAsync not implemented");
}

BOOL tty_isAsync(struct FsOpenNode* node, struct KProcess* process) {
    return 0;
}

void tty_waitForEvents(struct FsOpenNode* node, struct KThread* thread, U32 events) {
    //kpanic("tty_waitForEvents not implemented");
}

BOOL tty_isWriteReady(struct FsOpenNode* node) {
    return (node->flags & K_O_ACCMODE)!=K_O_RDONLY;
}

BOOL tty_isReadReady(struct FsOpenNode* node) {
    return 0;
}

U32 tty_map(struct Memory* memory, struct FsOpenNode* node, U32 address, U32 len, S32 prot, S32 flags, U64 off) {
    return 0;
}

BOOL tty_canMap(struct FsOpenNode* node) {
    return FALSE;
}

struct FsOpenNodeFunc ttyAccess = {tty_init, tty_length, tty_setLength, tty_getFilePointer, tty_seek, tty_read, tty_write, tty_close, tty_map, tty_canMap, tty_ioctl, tty_setAsync, tty_isAsync, tty_waitForEvents, tty_isWriteReady, tty_isReadReady};
